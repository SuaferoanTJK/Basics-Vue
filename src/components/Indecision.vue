<script>
export default {
  name: "Indecision",
  data() {
    return {
      question: null,
      answer: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Thinking...";
        const { answer } = await fetch("https://yesno.wtf/api").then(
          (response) => response.json()
        );
        this.answer = answer;
      } catch (error) {
        this.answer = "The API couldn't be loaded";
      }
    },
  },
  watch: {
    question(value) {
      this.isValidQuestion = false;
      if (!value.includes("?")) return;
      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
};
</script>

<template>
  <div class="w-full md:w-1/2 mx-auto mb-5">
    <div class="flex justify-center">
      <img src="../assets/FortuneTeller.png" alt="FT" class="w-80" />
    </div>
    <div class="flex flex-col space-y-10">
      <div class="space-y-2">
        <label for="question" class="font-bold text-white">Question</label>
        <input
          id="question"
          type="text"
          placeholder="Ask me anything"
          v-model="question"
          class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-gray-300 focus:shadow-none"
        />
        <p class="text-sm text-white">End with the question mark (?)</p>
      </div>
      <div
        v-show="isValidQuestion"
        class="text-center text-4xl space-y-5 font-semibold"
      >
        <h1 class="relative top-0 left-0 capitalize text-9xl text-white">
          {{ answer }}
        </h1>
      </div>
    </div>
  </div>
</template>
