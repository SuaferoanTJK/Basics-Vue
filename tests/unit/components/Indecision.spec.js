import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision";

describe("Indecision Unit Testing", () => {
  let component;

  // Mock fetch function
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeAll(() => {
    component = shallowMount(Indecision);
  });

  test("snapshot should matched", () => {
    expect(component.html()).toMatchSnapshot();
  });
  test("write in the input shouldn't trigger the getAnswer method", async () => {
    const getAnswerSpy = jest.spyOn(component.vm, "getAnswer");
    const input = component.find("input");
    await input.setValue("Hi world");
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });
  test("write the question mark (?) should trigger the getAnswer method", async () => {
    const getAnswerSpy = jest.spyOn(component.vm, "getAnswer");
    const input = component.find("input");
    await input.setValue("Hi world?");
    expect(getAnswerSpy).toHaveBeenCalledTimes(1);
  });
  test("getAnswer method", async () => {
    await component.vm.getAnswer();
    expect(component.vm.answer).toBe("yes");
  });
  test("getAnswer method - API Failed", async () => {
    fetch.mockImplementationOnce(() => Promise.reject("API is down"));
    await component.vm.getAnswer();
    expect(component.vm.answer).toBe("The API couldn't be loaded");
  });
});
