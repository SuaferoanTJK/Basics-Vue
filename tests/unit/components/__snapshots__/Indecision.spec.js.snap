// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Unit Testing snapshot should matched 1`] = `
<div class="w-full md:w-1/2 mx-auto mb-5">
  <div class="flex justify-center"><img src="" alt="FT" class="w-80"></div>
  <div class="flex flex-col space-y-10">
    <div class="space-y-2"><label for="question" class="font-bold text-white">Question</label><input id="question" type="text" placeholder="Ask me anything" class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-gray-300 focus:shadow-none">
      <p class="text-sm text-white">End with the question mark (?)</p>
    </div>
    <div class="text-center text-4xl space-y-5 font-semibold" style="display: none;">
      <h1 class="relative top-0 left-0 capitalize text-9xl text-white"></h1>
    </div>
  </div>
</div>
`;
